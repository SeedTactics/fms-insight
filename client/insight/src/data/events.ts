/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
import { addDays, addMonths, startOfMonth } from "date-fns";
import { PledgeStatus, Pledge, ActionBeforeMiddleware, PledgeToPromise } from "../store/middleware";
import { Vector } from "prelude-ts";
import { LazySeq } from "./lazyseq";

import * as api from "./api";
import * as cycles from "./events.cycles";
import * as matsummary from "./events.matsummary";
import * as simuse from "./events.simuse";
import * as inspection from "./events.inspection";
import { JobsBackend, LogBackend } from "./backend";

export enum AnalysisPeriod {
  Last30Days = "Last_30_Days",
  SpecificMonth = "Specific_Month",
}

export interface Last30Days {
  readonly latest_log_counter: number | undefined;
  readonly most_recent_10_events: Vector<Readonly<api.ILogEntry>>;

  readonly latest_scheduleId: string | undefined;

  readonly cycles: cycles.CycleState;
  readonly sim_use: simuse.SimUseState;
  readonly inspection: inspection.InspectionState;

  readonly mat_summary: matsummary.MatSummaryState;
}

export interface AnalysisMonth {
  readonly cycles: cycles.CycleState;
  readonly sim_use: simuse.SimUseState;
  readonly inspection: inspection.InspectionState;
}

const emptyAnalysisMonth: AnalysisMonth = {
  cycles: cycles.initial,
  sim_use: simuse.initial,
  inspection: inspection.initial,
};

export interface State {
  readonly loading_log_entries: boolean;
  readonly loading_job_history: boolean;
  readonly loading_error?: Error;

  readonly analysis_period: AnalysisPeriod;
  readonly analysis_period_month: Date;
  readonly loading_analysis_month_log: boolean;
  readonly loading_analysis_month_jobs: boolean;

  readonly last30: Last30Days;
  readonly selected_month: AnalysisMonth;
}

export const initial: State = {
  loading_log_entries: false,
  loading_job_history: false,
  analysis_period: AnalysisPeriod.Last30Days,
  analysis_period_month: startOfMonth(new Date()),
  loading_analysis_month_log: false,
  loading_analysis_month_jobs: false,

  last30: {
    latest_log_counter: undefined,
    latest_scheduleId: undefined,
    most_recent_10_events: Vector.empty(),
    cycles: cycles.initial,
    mat_summary: matsummary.initial,
    sim_use: simuse.initial,
    inspection: inspection.initial,
  },

  selected_month: emptyAnalysisMonth,
};

export enum ActionType {
  SetAnalysisLast30Days = "Events_SetAnalysisLast30Days",
  SetAnalysisMonth = "Events_SetAnalysisMonth",
  LoadRecentLogEntries = "Events_LoadRecentLogEntries",
  LoadRecentJobHistory = "Events_LoadRecentJobHistory",
  LoadSpecificMonthLogEntries = "Events_LoadSpecificMonthLogEntries",
  LoadSpecificMonthJobHistory = "Events_LoadSpecificMonthJobHistory",
  ReceiveNewLogEntries = "Events_NewLogEntries",
  ReceiveNewJobs = "Events_ReceiveNewJobs",
}

export type Action =
  | { type: ActionType.SetAnalysisLast30Days }
  | { type: ActionType.SetAnalysisMonth; month: Date }
  | {
      type: ActionType.LoadRecentLogEntries;
      now: Date;
      pledge: Pledge<ReadonlyArray<Readonly<api.ILogEntry>>>;
    }
  | {
      type: ActionType.LoadRecentJobHistory;
      now: Date;
      pledge: Pledge<Readonly<api.IHistoricData>>;
    }
  | {
      type: ActionType.LoadSpecificMonthLogEntries;
      month: Date;
      pledge: Pledge<ReadonlyArray<Readonly<api.ILogEntry>>>;
    }
  | {
      type: ActionType.LoadSpecificMonthJobHistory;
      month: Date;
      pledge: Pledge<Readonly<api.IHistoricData>>;
    }
  | {
      type: ActionType.ReceiveNewLogEntries;
      now: Date;
      events: ReadonlyArray<Readonly<api.ILogEntry>>;
    }
  | {
      type: ActionType.ReceiveNewJobs;
      now: Date;
      jobs: Readonly<api.IHistoricData>;
    };

type ABF = ActionBeforeMiddleware<Action>;

export function loadLast30Days(): ABF {
  const now = new Date();
  const thirtyDaysAgo = addDays(now, -30);
  return [
    {
      type: ActionType.LoadRecentLogEntries,
      now: now,
      pledge: LogBackend.get(thirtyDaysAgo, now),
    },
    {
      type: ActionType.LoadRecentJobHistory,
      now: now,
      pledge: JobsBackend.history(thirtyDaysAgo, now),
    },
  ];
}

export function refreshLogEntries(lastCounter: number): ABF {
  const now = new Date();
  return {
    type: ActionType.LoadRecentLogEntries,
    now: now,
    pledge: LogBackend.recent(lastCounter),
  };
}

export function receiveNewEvents(events: ReadonlyArray<Readonly<api.ILogEntry>>): ABF {
  return {
    type: ActionType.ReceiveNewLogEntries,
    now: new Date(),
    events,
  };
}

export function receiveNewJobs(newJobs: Readonly<api.INewJobs>): ABF {
  const jobs: { [key: string]: api.JobPlan } = {};
  newJobs.jobs.forEach((j) => {
    jobs[j.unique] = j;
  });
  return {
    type: ActionType.ReceiveNewJobs,
    now: new Date(),
    jobs: {
      jobs: jobs,
      stationUse: newJobs.stationUse || [],
    },
  };
}

export function analyzeLast30Days(): ABF {
  return { type: ActionType.SetAnalysisLast30Days };
}

export function analyzeSpecificMonth(month: Date): ReadonlyArray<PledgeToPromise<Action>> {
  const startOfNextMonth = addMonths(month, 1);
  return [
    {
      type: ActionType.LoadSpecificMonthLogEntries,
      month: month,
      pledge: LogBackend.get(month, startOfNextMonth),
    },
    {
      type: ActionType.LoadSpecificMonthJobHistory,
      month: month,
      pledge: JobsBackend.history(month, startOfNextMonth),
    },
  ];
}

export function setAnalysisMonth(month: Date): ABF {
  return {
    type: ActionType.SetAnalysisMonth,
    month: month,
  };
}

function safeAssign<T extends R, R>(o: T, n: R): T {
  let allMatch = true;
  for (const k of Object.keys(n)) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    if ((o as any)[k] !== (n as any)[k]) {
      allMatch = false;
      break;
    }
  }
  if (allMatch) {
    return o;
  } else {
    return Object.assign({}, o, n);
  }
}

function processRecentLogEntries(now: Date, evts: ReadonlyArray<Readonly<api.ILogEntry>>, s: Last30Days): Last30Days {
  const thirtyDaysAgo = addDays(now, -30);
  const initialLoad = s.latest_log_counter === undefined;
  let lastCounter = s.latest_log_counter;
  const lastNewEvent = LazySeq.ofIterable(evts).maxOn((e) => e.counter);
  let last10Evts = s.most_recent_10_events;
  if (lastNewEvent.isSome()) {
    if (lastCounter === undefined || lastCounter < lastNewEvent.get().counter) {
      lastCounter = lastNewEvent.get().counter;
    }
    const lastNew10 = evts.slice(-10);
    last10Evts = last10Evts.appendAll(lastNew10).reverse().take(10).reverse();
  }
  return safeAssign(s, {
    latest_log_counter: lastCounter,
    cycles: cycles.process_events(
      { type: cycles.ExpireOldDataType.ExpireEarlierThan, d: thirtyDaysAgo },
      evts,
      initialLoad,
      s.cycles
    ),
    most_recent_10_events: last10Evts,
    mat_summary: matsummary.process_events(now, evts, s.mat_summary),
    inspection: inspection.process_events(
      { type: cycles.ExpireOldDataType.ExpireEarlierThan, d: thirtyDaysAgo },
      evts,
      undefined,
      s.inspection
    ),
  });
}

function processSpecificMonthLogEntries(evts: ReadonlyArray<Readonly<api.ILogEntry>>, s: AnalysisMonth): AnalysisMonth {
  return safeAssign(s, {
    cycles: cycles.process_events(
      { type: cycles.ExpireOldDataType.NoExpire },
      evts,
      true, // initial load is true
      s.cycles
    ),
    inspection: inspection.process_events({ type: cycles.ExpireOldDataType.NoExpire }, evts, undefined, s.inspection),
  });
}

function processRecentJobs(now: Date, jobs: Readonly<api.IHistoricData>, s: Last30Days): Last30Days {
  const thirtyDaysAgo = addDays(now, -30);
  let latestSchId = s.latest_scheduleId;
  const largestSchId = LazySeq.ofObject(jobs.jobs)
    .maxOn(([_, j]) => j.scheduleId || "")
    .mapNullable(([_, j]) => j.scheduleId);
  if (largestSchId.isSome()) {
    if (latestSchId === undefined || latestSchId < largestSchId.get()) {
      latestSchId = largestSchId.get();
    }
  }

  return safeAssign(s, {
    latest_scheduleId: latestSchId,
    sim_use: simuse.process_sim_use(
      { type: cycles.ExpireOldDataType.ExpireEarlierThan, d: thirtyDaysAgo },
      jobs,
      s.sim_use
    ),
  });
}

function processSpecificMonthJobs(jobs: Readonly<api.IHistoricData>, s: AnalysisMonth): AnalysisMonth {
  return safeAssign(s, {
    sim_use: simuse.process_sim_use({ type: cycles.ExpireOldDataType.NoExpire }, jobs, s.sim_use),
  });
}

export function reducer(s: State | undefined, a: Action): State {
  if (s === undefined) {
    return initial;
  }
  switch (a.type) {
    case ActionType.LoadRecentLogEntries:
      switch (a.pledge.status) {
        case PledgeStatus.Starting:
          return { ...s, loading_log_entries: true, loading_error: undefined };
        case PledgeStatus.Completed:
          return {
            ...s,
            last30: processRecentLogEntries(a.now, a.pledge.result, s.last30),
            loading_log_entries: false,
          };
        case PledgeStatus.Error:
          return {
            ...s,
            loading_log_entries: false,
            loading_error: a.pledge.error,
          };
        default:
          return s;
      }

    case ActionType.ReceiveNewLogEntries:
      return {
        ...s,
        last30: processRecentLogEntries(a.now, a.events, s.last30),
      };

    case ActionType.LoadRecentJobHistory:
      switch (a.pledge.status) {
        case PledgeStatus.Starting:
          return { ...s, loading_job_history: true, loading_error: undefined };
        case PledgeStatus.Completed:
          return {
            ...s,
            last30: processRecentJobs(a.now, a.pledge.result, s.last30),
            loading_job_history: false,
          };
        case PledgeStatus.Error:
          return {
            ...s,
            loading_job_history: false,
            loading_error: a.pledge.error,
          };
        default:
          return s;
      }

    case ActionType.ReceiveNewJobs:
      return {
        ...s,
        last30: processRecentJobs(a.now, a.jobs, s.last30),
      };

    case ActionType.SetAnalysisLast30Days:
      return {
        ...s,
        analysis_period: AnalysisPeriod.Last30Days,
        selected_month: emptyAnalysisMonth,
      };

    case ActionType.LoadSpecificMonthLogEntries:
      switch (a.pledge.status) {
        case PledgeStatus.Starting:
          return {
            ...s,
            analysis_period: AnalysisPeriod.SpecificMonth,
            analysis_period_month: a.month,
            loading_error: undefined,
            loading_analysis_month_log: true,
            selected_month: emptyAnalysisMonth,
          };

        case PledgeStatus.Completed:
          return {
            ...s,
            loading_analysis_month_log: false,
            selected_month: processSpecificMonthLogEntries(a.pledge.result, s.selected_month),
          };

        case PledgeStatus.Error:
          return {
            ...s,
            loading_analysis_month_log: false,
            loading_error: a.pledge.error,
          };

        default:
          return s;
      }

    case ActionType.LoadSpecificMonthJobHistory:
      switch (a.pledge.status) {
        case PledgeStatus.Starting:
          return {
            ...s,
            loading_analysis_month_jobs: true,
          };

        case PledgeStatus.Completed:
          return {
            ...s,
            loading_analysis_month_jobs: false,
            selected_month: processSpecificMonthJobs(a.pledge.result, s.selected_month),
          };

        case PledgeStatus.Error:
          return {
            ...s,
            loading_analysis_month_jobs: false,
            loading_error: a.pledge.error,
          };

        default:
          return s;
      }

    case ActionType.SetAnalysisMonth:
      return { ...s, analysis_period_month: a.month };

    default:
      return s;
  }
}
