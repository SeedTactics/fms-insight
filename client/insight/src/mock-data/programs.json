[
  {
    "MachineGroupName": "Machine",
    "ProgramName": "100",
    "CellControllerProgramName": "100",
    "Revision": null,
    "Comment": "program 100 comment"
  },
  {
    "MachineGroupName": "Machine",
    "ProgramName": "200",
    "CellControllerProgramName": "200",
    "Revision": null,
    "Comment": "comment for 200"
  },
  {
    "MachineGroupName": "Machine",
    "ProgramName": "300",
    "Revision": null,
    "CellControllerProgramName": "300"
  },
  {
    "MachineGroupName": "Machine",
    "ProgramName": "400",
    "CellControllerProgramName": "400",
    "Revision": null
  }
]
