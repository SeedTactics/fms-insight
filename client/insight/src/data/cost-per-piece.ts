/* Copyright (c) 2019, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import { PartCycleData } from "./events.cycles";
import { getDaysInMonth } from "date-fns";
import { HashMap, Vector, HasEquals } from "prelude-ts";
import { LazySeq } from "./lazyseq";

export interface CostInput {
  // key is machine group name
  readonly machineCostPerYear: { readonly [key: string]: number };
  readonly partMaterialCost: { readonly [key: string]: number };
  readonly automationCostPerYear?: number;
  readonly numOperators: number;
  readonly operatorCostPerHour: number;
}

export enum ActionType {
  SetMachineCostPerYear = "CostPerPiece_SetMachineCost",
  SetPartMaterialCost = "CostPerPiece_SetPartCost",
  SetAutomationCost = "CostPerPiece_SetAutomationCost",
  SetNumOperators = "CostPerPiece_SetNumOpers",
  SetOperatorCostPerHour = "CostPerPiece_SetOperatorCost",
}

export type Action =
  | { type: ActionType.SetMachineCostPerYear; group: string; cost: number }
  | { type: ActionType.SetPartMaterialCost; part: string; cost: number }
  | { type: ActionType.SetAutomationCost; cost: number }
  | { type: ActionType.SetNumOperators; numOpers: number }
  | { type: ActionType.SetOperatorCostPerHour; cost: number };

export interface State {
  readonly input: CostInput;
}

export const initial: State = (function () {
  const json = localStorage.getItem("cost-per-piece");
  if (json) {
    return {
      input: JSON.parse(json),
    };
  } else {
    return {
      input: {
        machineCostPerYear: {},
        partMaterialCost: {},
        numOperators: 0,
        automationCostPerYear: 0,
        operatorCostPerHour: 0,
      },
    };
  }
})();

export function reducer(s: State, a: Action): State {
  if (s === undefined) {
    return initial;
  }
  let newSt = s;
  switch (a.type) {
    case ActionType.SetMachineCostPerYear:
      if (isNaN(a.cost)) {
        return s;
      }
      newSt = {
        ...s,
        input: {
          ...s.input,
          machineCostPerYear: {
            ...s.input.machineCostPerYear,
            [a.group]: a.cost,
          },
        },
      };
      break;
    case ActionType.SetPartMaterialCost:
      if (isNaN(a.cost)) {
        return s;
      }
      newSt = {
        ...s,
        input: {
          ...s.input,
          partMaterialCost: {
            ...s.input.partMaterialCost,
            [a.part]: a.cost,
          },
        },
      };
      break;
    case ActionType.SetAutomationCost:
      if (isNaN(a.cost)) {
        return s;
      }
      newSt = {
        ...s,
        input: {
          ...s.input,
          automationCostPerYear: a.cost,
        },
      };
      break;
    case ActionType.SetNumOperators:
      if (isNaN(a.numOpers)) {
        return s;
      }
      newSt = {
        ...s,
        input: {
          ...s.input,
          numOperators: a.numOpers,
        },
      };
      break;
    case ActionType.SetOperatorCostPerHour:
      if (isNaN(a.cost)) {
        return s;
      }
      newSt = {
        ...s,
        input: {
          ...s.input,
          operatorCostPerHour: a.cost,
        },
      };
      break;
  }
  if (s !== newSt) {
    localStorage.setItem("cost-per-piece", JSON.stringify(newSt.input));
  }
  return newSt;
}

export interface PartCost {
  readonly part: string;
  readonly parts_completed: number;

  // sum of machine costs over all cycles.  Must be divided by parts_completed to get cost/piece
  readonly machine_cost: number;

  // sum of labor costs over all cycles.  Must be divided by parts_completed to get cost/piece
  readonly labor_cost: number;

  // split of yearly automation cost over all use
  readonly automation_cost: number;

  readonly material_cost: number;
}

function machine_cost(
  cycles: LazySeq<PartCycleData>,
  totalStatUseMinutes: HashMap<string, number>,
  machineCostPerYear: { readonly [key: string]: number },
  days: number
): number {
  return cycles
    .toMap(
      (c) => [c.stationGroup, c.activeMinutes],
      (a1, a2) => a1 + a2
    )
    .foldLeft(0, (x: number, [statGroup, minutes]: [string, number]) => {
      const totalUse = totalStatUseMinutes.get(statGroup).getOrElse(1);
      const totalMachineCost = ((machineCostPerYear[statGroup] || 0) * days) / 365;
      return x + (minutes / totalUse) * totalMachineCost;
    });
}

function labor_cost(
  cycles: LazySeq<PartCycleData>,
  totalStatUseMinutes: HashMap<string, number>,
  totalLaborCost: number
): number {
  const pctUse = cycles
    .toMap(
      (c) => [c.stationGroup, c.activeMinutes],
      (a1, a2) => a1 + a2
    )
    .foldLeft(0, (x: number, [statGroup, minutes]: [string, number]) => {
      const total = totalStatUseMinutes.get(statGroup).getOrElse(1);
      return x + minutes / total;
    });
  return pctUse * totalLaborCost;
}

function auto_cost(
  cycles: LazySeq<PartCycleData>,
  totalPalletCycles: number,
  autoCostPerYear: number | undefined
): number {
  if (!autoCostPerYear) {
    return 0;
  }
  const pctUse = cycles.sumOn((v) => (v.completed ? 1 : 0)) / totalPalletCycles;
  return (pctUse * autoCostPerYear) / 12;
}

export function compute_monthly_cost(
  i: CostInput,
  cycles: Vector<PartCycleData>,
  month?: Date
): ReadonlyArray<PartCost> {
  const days = month ? getDaysInMonth(month) : 30;
  const totalLaborCost = days * 24 * i.operatorCostPerHour * i.numOperators;

  const totalStatUseMinutes: HashMap<string, number> = LazySeq.ofIterable(cycles).toMap(
    (c) => [c.stationGroup, c.activeMinutes],
    (a1, a2) => a1 + a2
  );

  const totalPalletCycles = LazySeq.ofIterable(cycles).sumOn((v) => (v.completed ? 1 : 0));

  return Array.from(
    LazySeq.ofIterable(cycles)
      .groupBy((c) => c.part)
      .map((partName, forPart) => [
        partName as string & HasEquals,
        {
          part: partName,
          parts_completed: LazySeq.ofIterable(forPart)
            .filter((c) => c.completed)
            .sumOn((c) => c.material.length),
          machine_cost: machine_cost(
            LazySeq.ofIterable(forPart).filter((c) => !c.isLabor),
            totalStatUseMinutes,
            i.machineCostPerYear,
            days
          ),
          labor_cost: labor_cost(
            LazySeq.ofIterable(forPart).filter((c) => c.isLabor),
            totalStatUseMinutes,
            totalLaborCost
          ),
          automation_cost: auto_cost(LazySeq.ofIterable(forPart), totalPalletCycles, i.automationCostPerYear),
          material_cost: i.partMaterialCost[partName] || 0,
        },
      ])
      .valueIterable()
  );
}
