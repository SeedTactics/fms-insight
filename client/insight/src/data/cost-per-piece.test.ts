/* Copyright (c) 2019, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import * as ccp from "./cost-per-piece";
import * as cycles from "./events.cycles";
import { loadMockData } from "../mock-data/load";

it("loads the initial state", () => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const s = ccp.reducer(undefined as any, undefined as any);
  expect(s).toBe(ccp.initial);
});

it("sets machine costs", () => {
  let s = ccp.reducer(ccp.initial, {
    type: ccp.ActionType.SetMachineCostPerYear,
    group: "g1",
    cost: 12,
  });
  expect(s).toEqual({
    input: {
      ...ccp.initial.input,
      machineCostPerYear: {
        g1: 12,
      },
    },
  });

  s = ccp.reducer(s, {
    type: ccp.ActionType.SetMachineCostPerYear,
    group: "g2",
    cost: 44,
  });
  expect(s).toEqual({
    input: {
      ...ccp.initial.input,
      machineCostPerYear: {
        g1: 12,
        g2: 44,
      },
    },
  });

  s = ccp.reducer(s, {
    type: ccp.ActionType.SetMachineCostPerYear,
    group: "g2",
    cost: 150,
  });
  expect(s).toEqual({
    input: {
      ...ccp.initial.input,
      machineCostPerYear: {
        g1: 12,
        g2: 150,
      },
    },
  });
  expect(JSON.parse(localStorage.getItem("cost-per-piece") || "")).toEqual({
    ...ccp.initial.input,
    machineCostPerYear: {
      g1: 12,
      g2: 150,
    },
  });
});

it("sets part material cost", () => {
  let s = ccp.reducer(ccp.initial, {
    type: ccp.ActionType.SetPartMaterialCost,
    part: "p1",
    cost: 52,
  });
  expect(s).toEqual({
    input: {
      ...ccp.initial.input,
      partMaterialCost: {
        p1: 52,
      },
    },
  });

  s = ccp.reducer(s, {
    type: ccp.ActionType.SetPartMaterialCost,
    part: "p2",
    cost: 2,
  });
  expect(s).toEqual({
    input: {
      ...ccp.initial.input,
      partMaterialCost: {
        p1: 52,
        p2: 2,
      },
    },
  });

  s = ccp.reducer(s, {
    type: ccp.ActionType.SetPartMaterialCost,
    part: "p2",
    cost: 63,
  });
  expect(s).toEqual({
    input: {
      ...ccp.initial.input,
      partMaterialCost: {
        p1: 52,
        p2: 63,
      },
    },
  });
  expect(JSON.parse(localStorage.getItem("cost-per-piece") || "")).toEqual({
    ...ccp.initial.input,
    partMaterialCost: {
      p1: 52,
      p2: 63,
    },
  });
});

it("sets the number of operators", () => {
  const s = ccp.reducer(ccp.initial, {
    type: ccp.ActionType.SetNumOperators,
    numOpers: 12,
  });
  expect(s).toEqual({
    input: {
      ...ccp.initial.input,
      numOperators: 12,
    },
  });
  expect(JSON.parse(localStorage.getItem("cost-per-piece") || "")).toEqual({
    ...ccp.initial.input,
    numOperators: 12,
  });
});

it("sets the operator cost per hour", () => {
  const s = ccp.reducer(ccp.initial, {
    type: ccp.ActionType.SetOperatorCostPerHour,
    cost: 67,
  });
  expect(s).toEqual({
    input: {
      ...ccp.initial.input,
      operatorCostPerHour: 67,
    },
  });
  expect(JSON.parse(localStorage.getItem("cost-per-piece") || "")).toEqual({
    ...ccp.initial.input,
    operatorCostPerHour: 67,
  });
});

it("sets the automation cost per year", () => {
  const s = ccp.reducer(ccp.initial, {
    type: ccp.ActionType.SetAutomationCost,
    cost: 123456,
  });
  expect(s).toEqual({
    input: {
      ...ccp.initial.input,
      automationCostPerYear: 123456,
    },
  });
  expect(JSON.parse(localStorage.getItem("cost-per-piece") || "")).toEqual({
    ...ccp.initial.input,
    automationCostPerYear: 123456,
  });
});

it("computes part costs", async () => {
  const data = loadMockData(30 * 24 * 60 * 60);
  const evts = await data.events;
  const cycleState = cycles.process_events({ type: cycles.ExpireOldDataType.NoExpire }, evts, true, cycles.initial);
  const costInput: ccp.CostInput = {
    machineCostPerYear: {
      Machine: 1234,
    },
    partMaterialCost: {
      aaa: 15,
      bbb: 19,
    },
    numOperators: 2,
    automationCostPerYear: 10000,
    operatorCostPerHour: 5432,
  };

  expect(ccp.compute_monthly_cost(costInput, cycleState.part_cycles)).toMatchSnapshot("cost calcs");
});
