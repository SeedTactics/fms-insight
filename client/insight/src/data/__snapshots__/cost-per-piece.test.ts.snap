// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computes part costs: cost calcs 1`] = `
Array [
  Object {
    "automation_cost": 821.917808219178,
    "labor_cost": 1298080.9620340148,
    "machine_cost": 24.924370475942847,
    "part": "aaa",
    "parts_completed": 240,
  },
  Object {
    "automation_cost": 821.917808219178,
    "labor_cost": 1655523.5457825116,
    "machine_cost": 35.82595186816961,
    "part": "bbb",
    "parts_completed": 307,
  },
  Object {
    "automation_cost": 821.917808219178,
    "labor_cost": 2064029.3557807938,
    "machine_cost": 102.69565292984005,
    "part": "ccc",
    "parts_completed": 383,
  },
  Object {
    "automation_cost": 821.917808219178,
    "labor_cost": 1083077.904140182,
    "machine_cost": 53.19519293243262,
    "part": "xxx",
    "parts_completed": 402,
  },
  Object {
    "automation_cost": 821.917808219178,
    "labor_cost": 1056202.521903453,
    "machine_cost": 71.2062591214296,
    "part": "yyy",
    "parts_completed": 392,
  },
  Object {
    "automation_cost": 821.917808219178,
    "labor_cost": 665165.7103590448,
    "machine_cost": 117.85120280917151,
    "part": "zzz",
    "parts_completed": 247,
  },
]
`;

exports[`computes part costs: cost clipboard table 1`] = `
<div>
  <table>
    

    <thead>
      <tr>
        <th>
          Part
        </th>
        <th>
          Completed Quantity
        </th>
        <th>
          Machine Cost
        </th>
        <th>
          Labor Cost
        </th>
        <th>
          Automation Cost
        </th>
        <th>
          Total
        </th>
      </tr>
    </thead>
    

    <tbody>
      

      <tr>
        <td>
          aaa
        </td>
        <td>
          240
        </td>
        <td>
          0.1
        </td>
        <td>
          5,408.7
        </td>
        <td>
          3.4
        </td>
        <td>
          5,412.2
        </td>
      </tr>
      

      <tr>
        <td>
          bbb
        </td>
        <td>
          307
        </td>
        <td>
          0.1
        </td>
        <td>
          5,392.6
        </td>
        <td>
          2.7
        </td>
        <td>
          5,395.4
        </td>
      </tr>
      

      <tr>
        <td>
          ccc
        </td>
        <td>
          383
        </td>
        <td>
          0.3
        </td>
        <td>
          5,389.1
        </td>
        <td>
          2.1
        </td>
        <td>
          5,391.5
        </td>
      </tr>
      

      <tr>
        <td>
          xxx
        </td>
        <td>
          402
        </td>
        <td>
          0.1
        </td>
        <td>
          2,694.2
        </td>
        <td>
          2
        </td>
        <td>
          2,696.4
        </td>
      </tr>
      

      <tr>
        <td>
          yyy
        </td>
        <td>
          392
        </td>
        <td>
          0.2
        </td>
        <td>
          2,694.4
        </td>
        <td>
          2.1
        </td>
        <td>
          2,696.7
        </td>
      </tr>
      

      <tr>
        <td>
          zzz
        </td>
        <td>
          247
        </td>
        <td>
          0.5
        </td>
        <td>
          2,693
        </td>
        <td>
          3.3
        </td>
        <td>
          2,696.8
        </td>
      </tr>
      

    </tbody>
    

  </table>
</div>
`;
